### PR TITLE
[JSC] Don't use JSScope instances as `this` value when calling functions

### PR DESCRIPTION
**Pull request by shvaikalesh**
#### ee167b8fe4fd234a33b2381640cba982fa6c7516
<pre>
[JSC] Don&apos;t use JSScope instances as `this` value when calling functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=225397">https://bugs.webkit.org/show_bug.cgi?id=225397</a>
&lt;<a href="https://rdar.apple.com/problem/77915872">rdar://problem/77915872</a>&gt;

Reviewed by Yusuke Suzuki.

Over the past few years, we&apos;ve been constantly encountering security vulnerabilities / spec compat issues
that arise due to necessity of `ToThis` even in strict mode (which is a mode for all built-ins):

  <a href="https://commits.webkit.org/229319@main">https://commits.webkit.org/229319@main</a>
  <a href="https://commits.webkit.org/259016@main">https://commits.webkit.org/259016@main</a>
  <a href="https://commits.webkit.org/272163@main">https://commits.webkit.org/272163@main</a> (security)
  <a href="https://commits.webkit.org/276104@main">https://commits.webkit.org/276104@main</a> (security)
  <a href="https://github.com/WebKit/WebKit/pull/22329">https://github.com/WebKit/WebKit/pull/22329</a> (caught during review)

`ToThis` in strict mode is very easy to miss because it doesn&apos;t make sense in terms of ECMA-262 spec.
Passing the resolved scope as `this` value is a sorta of hack to support `with` statement.

This change replaces the above-mentioned footgun with a `JSType` check of the resolved scope,
which is performed only inside `with` statement by making use of `TaintedByWithScopeLexicallyScopedFeature`,
and dramatically simplifies `ToThis`, which is now emitted only in sloppy mode.

The downside of this approach is that we need to ensure that every `JSC::call()` and `JSC::evaluate()`
is called with `JSGlobalProxy` as `this` value rather than raw global object, which is now enforced
by assertions and implemented with `wrapGlobalObject` helper.

Also, this change fixes minor ECMA-262 and WebIDL spec compat issues, like calling cross-realm getters
accessing `__proto__` from the scope.

This patch reduces 1 bytecode of every strict function, but adds 1 bytecode for every call from
scope, yet was proven to be neutral on JetStream3 and Speedometer3.

* JSTests/stress/big-int-strict-spec-to-this.js:
* JSTests/stress/evaluate-with-scope-extension.js: Added.
* JSTests/stress/global-object-accessor-property-this-value.js: Added.
* JSTests/stress/variable-named-eval.js: Added.
* LayoutTests/fast/dom/Window/dom-access-from-closure-iframe-expected.txt:
* LayoutTests/fast/dom/Window/dom-access-from-closure-window-expected.txt:
* LayoutTests/fast/dom/Window/dom-access-from-closure-window-with-gc-expected.txt:
* LayoutTests/http/tests/security/xss-DENIED-htmlelelment-with-iframe-proto-expected.txt:
* LayoutTests/http/tests/security/xss-DENIED-method-with-iframe-proto-expected.txt:
* LayoutTests/http/tests/security/xss-DENIED-non-shadowable-propterty-with-iframe-proto-expected.txt:
* LayoutTests/http/tests/security/xss-DENIED-regular-propterty-with-iframe-proto-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/global-object-implicit-this-value-cross-realm-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/global-object-implicit-this-value.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/global-object-implicit-this-value.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/global-object-implicit-this-value.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/global-object-implicit-this-value.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/windows/embedded-opener-remove-frame-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/execution-timing/083-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/websockets/unload-a-document/001_default-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/websockets/unload-a-document/001_wss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/websockets/unload-a-document/002_default-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/websockets/unload-a-document/002_wss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/websockets/unload-a-document/003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/websockets/unload-a-document/004-expected.txt:
* LayoutTests/js/object-literal-shorthand-construction-expected.txt:
* LayoutTests/js/script-tests/object-literal-shorthand-construction.js:
(set 2):
(get 1):
* LayoutTests/storage/indexeddb/resources/crash-on-getdatabases.js:
(async testDoesNotCrash):
* Source/JavaScriptCore/API/APICallbackFunction.h:
(JSC::APICallbackFunction::callImpl):
* Source/JavaScriptCore/API/JSBase.cpp:
(JSEvaluateScriptInternal):
* Source/JavaScriptCore/API/JSCallbackObjectFunctions.h:
(JSC::JSCallbackObject&lt;Parent&gt;::callImpl):
* Source/JavaScriptCore/API/JSContextRef.cpp:
(JSContextGetGlobalObject):
* Source/JavaScriptCore/API/JSObjectRef.cpp:
(JSObjectCallAsFunction):
* Source/JavaScriptCore/API/JSScriptRef.cpp:
* Source/JavaScriptCore/API/tests/testapi.c:
(main):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/ProxyHelpers.js:
(linkTimeConstant.performProxyObjectGet):
(linkTimeConstant.performProxyObjectGetByVal):
(linkTimeConstant.performProxyObjectSetSloppy):
(linkTimeConstant.performProxyObjectSetStrict):
(linkTimeConstant.performProxyObjectSetByValSloppy):
(linkTimeConstant.performProxyObjectSetByValStrict):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finishCreation):
(JSC::CodeBlock::finalizeLLIntInlineCaches):
* Source/JavaScriptCore/bytecode/SpeculatedType.cpp:
(JSC::speculationFromJSTypeRange):
* Source/JavaScriptCore/bytecode/SpeculatedType.h:
* Source/JavaScriptCore/bytecode/ToThisStatus.cpp: Removed.
* Source/JavaScriptCore/bytecode/ToThisStatus.h: Removed.
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitEqualityOpImpl):
(JSC::BytecodeGenerator::emitGetFunctionFromScope):
(JSC::BytecodeGenerator::isTaintedByWithScope const):
(JSC::BytecodeGenerator::emitIsCellWithType):
(JSC::BytecodeGenerator::emitToThis):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitToThis): Deleted.
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::TaggedTemplateNode::emitBytecode):
(JSC::EvalFunctionCallNode::emitBytecode):
(JSC::FunctionCallResolveNode::emitBytecode):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_toThis): Deleted.
* Source/JavaScriptCore/debugger/DebuggerCallFrame.cpp:
(JSC::DebuggerCallFrame::thisValue const):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
(JSC::DFG::isToThisAnIdentity): Deleted.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
(JSC::DFG::ByteCodeParser::handleIntrinsicGetter):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupToThis):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::queriedType):
(JSC::DFG::Node::speculatedTypeForQuery):
(JSC::DFG::Node::hasECMAMode):
(JSC::DFG::Node::ecmaMode):
(JSC::DFG::Node::hasSpeculatedTypeForQuery): Deleted.
* Source/JavaScriptCore/dfg/DFGOpInfo.h:
(JSC::DFG::OpInfo::OpInfo):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
(JSC::DFG::putDynamicVar):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileToThis):
* Source/JavaScriptCore/generator/DSL.rb:
* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::JSInjectedScriptHost::internalConstructorName):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeCallImpl):
(JSC::Interpreter::executeEval):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::isCellWithType):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_is_cell_with_type):
(JSC::JIT::emit_op_to_this):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::createArrayIteratorObject):
* Source/JavaScriptCore/runtime/AsyncIteratorPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::evaluate):
(JSC::evaluateWithScopeExtension):
* Source/JavaScriptCore/runtime/DatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ErrorPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSCJSValue.cpp:
(JSC::JSValue::toThisSloppySlowCase const): Deleted.
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::toSloppyModeThis const):
(JSC::JSValue::toThis const): Deleted.
* Source/JavaScriptCore/runtime/JSCast.h:
(JSC::JSTypeRange::rawBits const):
(JSC::JSTypeRange::dump const):
(JSC::JSTypeRange::fromBits):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::wrapGlobalObject):
(JSC::JSScope::globalThis): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::setPrototypeWithCycleCheck):
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::JSObject::isStrictEvalActivation const): Deleted.
(JSC::JSObject::isEnvironment const): Deleted.
* Source/JavaScriptCore/runtime/JSScope.h:
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/ObjectPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::performProxyGet):
(JSC::ProxyObject::performPut):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/SparseArrayValueMap.cpp:
(JSC::SparseArrayEntry::put):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::checkObjectCoercible):
* Source/WebCore/bindings/js/JSDOMOperation.h:
(WebCore::IDLOperation::call):
(WebCore::IDLOperation::cast): Deleted.
* Source/WebCore/bindings/js/JSDOMOperationReturningPromise.h:
(WebCore::IDLOperationReturningPromise::call):
(WebCore::IDLOperationReturningPromise::callReturningPromisePair):
(WebCore::IDLOperationReturningPromise::callReturningOwnPromise):
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSErrorHandler.cpp:
(WebCore::JSErrorHandler::handleEvent):
* Source/WebCore/bindings/js/JSEventTargetCustom.h:
(WebCore::IDLOperation&lt;JSEventTarget&gt;::call):
* Source/WebCore/bindings/js/ScheduledAction.cpp:
(WebCore::ScheduledAction::execute):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateAttributeSetterBodyDefinition):
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_putForwardsAttributeGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::setJSTestGlobalObject_putForwardsAttributeSetter):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
* Source/WebCore/bindings/scripts/test/TestGlobalObject.idl:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::ensureMediaControls):

Canonical link: <a href="https://commits.webkit.org/284931@main">https://commits.webkit.org/284931@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/20dd7d612e9eddca36323475c75a0874d2635a7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/70939 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/50351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/23712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/22144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/73055 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/58150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/21962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/75043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/22144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/74005 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/58150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/23712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/58150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/23712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/20485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/64063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/58150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/23712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/76758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/70188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/15171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/21962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/76758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/15215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/23712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/76758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/23712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/91969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10881 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/46152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/91969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/47224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/48507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/46966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/JavaScriptCore/dfg/DFGClobberize.h [hunk 1]
```diff
@@ -797,6 +791,7 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
     case ResolveScopeForHoistingFuncDeclInEval:
     case ResolveScope:
     case ToObject:
+    case ToThis:
```
Comment (Constellation) [inline Source/JavaScriptCore/dfg/DFGClobberize.h]: Why does `ToThis` become much more conservative here? (clobber top)

File: Source/JavaScriptCore/jit/JITOpcodes.cpp [hunk 2]
```diff
@@ -1652,6 +1637,29 @@ void JIT::emit_op_create_this(const JSInstruction* currentInstruction)
...
+{
+    auto bytecode = currentInstruction->as<OpToThis>();
+    emitGetVirtualRegister(bytecode.m_srcDst, jsRegT10);
+
+    Jump isCell = branchIfCell(jsRegT10);
+    addSlowCase(branchIfNotOther(jsRegT10, regT3));
+
+#if USE(JSVALUE32_64)
+    move(TrustedImm32(JSValue::CellTag), jsRegT10.tagGPR());
+#endif
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/JITOpcodes.cpp]: We should do `boxCell` before setting `emitPutVirtualRegister`.

File: JSTests/stress/big-int-spec-to-this.js [hunk 3]
```diff
@@ -18,6 +17,3 @@ for (let i = 0; i < 10000; i++) {
 for (let i = 0; i < 10000; i++) {
     assert(foo.apply(300) === "object");
 }
-
-assert(numberOfDFGCompiles(foo) === 1);
-
```
Comment (MenloDorian) [inline JSTests/stress/big-int-spec-to-this.js]: Can you explain why this is assert is invalid?  Why did you remove it?

File: JSTests/stress/big-int-strict-spec-to-this.js [hunk 4]
```diff
@@ -18,6 +18,3 @@ for (let i = 0; i < 10000; i++) {
 for (let i = 0; i < 10000; i++) {
     assert(foo.apply(300) === "number");
 }
-
-assert(numberOfDFGCompiles(foo) > 1);
-
```
Comment (MenloDorian) [inline JSTests/stress/big-int-strict-spec-to-this.js]: Ditto.  Why remove?

File: JSTests/stress/big-int-spec-to-this.js (hunk 3 above)
Comment (shvaikalesh) [reply JSTests/stress/big-int-spec-to-this.js]: Oh I guess I was seeing flaky results locally because I missed the `useConcurrentJIT=0`. Brought back.

File: JSTests/stress/big-int-strict-spec-to-this.js (hunk 4 above)
Comment (shvaikalesh) [reply JSTests/stress/big-int-strict-spec-to-this.js]: Refined instead of removing: `ToThis` is now simpler, so it takes only 1 DFG compilation.

File: Source/JavaScriptCore/dfg/DFGClobberize.h (hunk 1 above)
Comment (shvaikalesh) [reply Source/JavaScriptCore/dfg/DFGClobberize.h]: Reverted, and adjusted AI.

File: Source/JavaScriptCore/jit/JITOpcodes.cpp (hunk 2 above)
Comment (shvaikalesh) [reply Source/JavaScriptCore/jit/JITOpcodes.cpp]: Done.

**Review by Constellation — APPROVED** [review]
r=me
